src/Raven.Server/Documents/Sharding/Background/ShardedDocumentsMigrator.cs

```diff
@@ -0,0 +1,106 @@
+
+                    Monitor.TryEnter(this, 0, ref monitorTaken);
+                    if (monitorTaken == false)
+                        return;
+
+                    if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                        return;
+
+                    using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                    using (context.OpenReadTransaction())
+                    {
+                        var configuration = _database.ShardingConfiguration;
+                        for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                        {
+                            var range = configuration.BucketRanges[index];
+                            if (range.ShardNumber == _database.ShardNumber)
+                                continue;
+
+                            var start = range.BucketRangeStart;
+                            var end = index == configuration.BucketRanges.Count - 1
+                                ? int.MaxValue
+                                : configuration.BucketRanges[index + 1].BucketRangeStart - 1;
+
+                            var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                            foreach (var bucketStats in bucketStatistics)
+                            {
+                                if (bucketStats.NumberOfDocuments == 0)
+                                    continue;
+
+                                bucket = bucketStats.Bucket;
+                                moveToShard = range.ShardNumber;
+                                found = true;
+                                break;
+                            }
+
+                            if (found)
+                                break;
+                        }
+                    }
+                }
+                finally
+                {
+                    if (monitorTaken)
+                        Monitor.Exit(this);
+                }
+
+                if (found)
+                    await MoveDocumentsToShardAsync(bucket, moveToShard);
+            }
+            catch (Exception e)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+
+                if (_logger.IsOperationsEnabled)
+                    _logger.Operations($"Failed to execute documents migration for '{_database.Name}'", e);
+
+                throw;
+            }
+        }
+
+        private async Task MoveDocumentsToShardAsync(int bucket, int moveToShard)
+        {
+            var cmd = new StartBucketMigrationCommand(bucket, _database.ShardNumber, moveToShard, _database.ShardedDatabaseName,
+                $"{Guid.NewGuid()}/{bucket}");
+
+            await _database.ServerStore.SendToLeaderAsync(cmd);
+        }
+
+        public void Dispose() => Monitor.Enter(this);
```

Comment:
That would leave this lock hanging.
It is okay now (in the sense that this will stall until this is done `ExecuteMoveDocumentsAsync` is done running, since we are not trying to wait on the lock.
But it is dangerous if we make changes in the future.

Also, if `ExecuteMoveDocumentsAsync` scheduled work, do we need to actually wait for it to complete?